### PR TITLE
Create `User` and `Breadcrumb` from map

### LINKED ISSUE #2687: Create `User` and `Breadcrumb` data classes from a dictionary
### Description

We have a lot of duplicate code across Hybrid SDKs, as most of them need to create the native data classes from a map.

- https://github.com/getsentry/team-mobile/issues/59

The goal is to add the following:
`[[SentryBreadcrumb alloc] initWithDict:dictionary]`
`[[SentryUser alloc] initWithDict:dictionary]`

As all the hybrid SDKs can use this.

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

Expose initialiser to create `User` and `Breadcrumb` from dictionary data.

## :bulb: Motivation and Context

Relates to https://github.com/getsentry/team-mobile/issues/59

Closes #2687

## :green_heart: How did you test it?

Unit tests

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Unreleased
 
+### Features
+
+- Create User and Breadcrumb from map #2820
```
Comment (brustolin) [inline CHANGELOG.md:7]:
```suggestion
- Create User and Breadcrumb from map (#2820)
```

File: Sources/Sentry/PrivateSentrySDKOnly.m [hunk 2]
```diff
@@ -134,4 +138,14 @@ + (NSData *)captureViewHierarchy
...
 
++ (SentryUser *_Nullable)userWithJSONObject:(NSDictionary *)jsonObject
+{
+    return [[SentryUser alloc] initWithJSONObject:jsonObject];
+}
+
++ (SentryBreadcrumb *_Nullable)breadcrumbWithJSONObject:(NSDictionary *)jsonObject
+{
+    return [[SentryBreadcrumb alloc] initWithJSONObject:jsonObject];
+}
```
Comment (brustolin) [inline Sources/Sentry/PrivateSentrySDKOnly.m:149]: m: To keep the pattern I suggest to ditch the "JSONObject" and use Dictionary.
And same for the actual implementation.

File: Sources/Sentry/include/HybridPublic/SentryBreadcrumb+Json.h
Comment (brustolin) [inline Sources/Sentry/include/HybridPublic/SentryBreadcrumb+Json.h:1]: h: I think this should be called SentryBreadCrumb+Private.h to keep the pattern.
Same for SentryUser+Json.h

**Review by brustolin — COMMENTED** [review]
First look, a few things I believe should be changed.

**Review by brustolin — APPROVED** [review]
LGTM

File: Sources/Sentry/SentryBreadcrumb.m [hunk 3]
```diff
@@ -3,8 +3,48 @@
...
+{
+    if (self = [super init]) {
+        NSMutableDictionary *unknown = [NSMutableDictionary dictionary];
+        for (id key in dictionary) {
+            id value = [dictionary valueForKey:key];
+            if (value == nil) {
+                continue;
+            }
+            Boolean isString = [value isKindOfClass:[NSString class]];
+            Boolean isDictionary = [value isKindOfClass:[NSDictionary class]];
```
Comment (philipphofmann) [inline Sources/Sentry/SentryBreadcrumb.m:23]: `l`:
```suggestion
            BOOL isString = [value isKindOfClass:[NSString class]];
            BOOL isDictionary = [value isKindOfClass:[NSDictionary class]];
```

File: Sources/Sentry/SentryUser.m [hunk 4]
```diff
@@ -3,8 +3,48 @@
...
+{
+    if (self = [super init]) {
+        NSMutableDictionary *unknown = [NSMutableDictionary dictionary];
+        for (id key in dictionary) {
+            id value = [dictionary valueForKey:key];
+            if (value == nil) {
+                continue;
+            }
+            Boolean isString = [value isKindOfClass:[NSString class]];
+            Boolean isDictionary = [value isKindOfClass:[NSDictionary class]];
```
Comment (philipphofmann) [inline Sources/Sentry/SentryUser.m:23]: `l`:
```suggestion
            BOOL isString = [value isKindOfClass:[NSString class]];
            BOOL isDictionary = [value isKindOfClass:[NSDictionary class]];
```

**Review by philipphofmann — COMMENTED** [review]
Just gave this a glance as @brustolin already approved.

**Comment by denrase** [comment]
@philipphofmann Having trouble to satisfy CI as an import of `SentryUser.h` or `SentryBreadcrumb.h` is not found. I guess that it ha to be additionally imported in another header. Could you point me to the correct place?

**Comment by brustolin** [comment]
@denrase, I fixed it for you

**Comment by brustolin** [comment]
Still need to fix other tests not compiling.

File: Sources/Sentry/SentryBreadcrumb.m [hunk 5]
```diff
@@ -3,8 +3,48 @@
...
 #import "SentryLevelMapper.h"
 
+@interface
+SentryBreadcrumb ()
+@property (atomic, strong) NSDictionary<NSString *, id> *_Nullable unknown;
+@end
+
 @implementation SentryBreadcrumb
 
+- (nullable instancetype)initWithDictionary:(NSDictionary *_Nullable)dictionary
```
Comment (armcknight) [inline Sources/Sentry/SentryBreadcrumb.m:13]: I wonder about the `nullable` here... of course, `[super init]` could always return `nil` in certain cases, but usually I see `init` methods declared in nonnull regions, so unless this method would explicitly return `nil` for failure to validate the input dictionary, maybe this and the user init should be nonnull instead.

**Review by armcknight — COMMENTED** [review]
Just a couple small things from me.

File: Sources/Sentry/SentryBreadcrumb.m (hunk 5 above)
Comment (denrase) [reply Sources/Sentry/SentryBreadcrumb.m:13]: We assumed that we might return nil if there is an error in the dererialization, but this is not the case with the current implementation. I will update the API to refelct that. Also, passing a nil dictionary also does not make sense as an API.

**Comment by denrase** [comment]
@brustolin @armcknight Merged in main and fixed changelog. Think we can merge?

**Pull request merged**